rm/rm-server/src/test/java/functionaltests/nodesource/TestNodeSourceThreadPool.java

```diff
@@ -98,12 +103,69 @@ public void testDownNodesCanStillBeDetectedWhenDeploymentThreadsAreExhausted() t
                                                            monitor);
 
         assertEquals(NodeState.DOWN, nodeEvent.getNodeState());
     }
 
+    @Test
+    /*
+     * This setup makes LOST nodes on one host and FREE nodes on another host.
+     * We use a configuration of two threads for the nodes deployment, and an
+     * infrastructure with infinite retry number for deploying. After cycling
+     * two redeployment phases thanks to the policy, we expect the situation
+     * to be:
+     * - one thread being always taken by the infinite retry on the fake host
+     * - one thread available to deploy with success/redeploy on the localhost
+     * A timeout is thrown if a thread leak prevents the redeployment to happen
+     */
+    public void testNodeSourceDeploymentThreadsDoNotLeakWithInfiniteRetry() throws Exception {
+        String rmConfig = new File(RMTHelper.class.getResource(SIX_THREADS_CONFIG_FILE_PATH).toURI()).getAbsolutePath();
+        this.rmHelper.startRM(rmConfig);
+        this.resourceManager = this.rmHelper.getResourceManager();
+        this.resourceManager.defineNodeSource(NODE_SOURCE_NAME,
+                                              SSHInfrastructureV2.class.getName(),
+                                              getInfiniteRetryInfrastructureParameters(),
+                                              RestartDownNodesPolicy.class.getName(),
+                                              getRestartDownNodesPolicyParameters(),
+                                              NODES_NOT_RECOVERABLE);
+        this.resourceManager.deployNodeSource(NODE_SOURCE_NAME);
+
+        RMTHelper.log("Waiting for the RM to have one free node");
+        while (this.resourceManager.getState().getFreeNodesNumber() != ONE_NODE_PER_HOST) {
+            Thread.sleep(ACTIVE_WAITING_PERIOD);
+        }
+
+        assertEquals(1, this.resourceManager.getState().getAliveNodes().size());
+
+        for (int i = 0; i < 2; i++) {
+            String freeNodeUrl = this.resourceManager.getState().getFreeNodes().iterator().next();
+            RMTHelper.killRuntime(freeNodeUrl);
+            RMMonitorsHandler monitor = this.rmHelper.getMonitorsHandler();
+            monitor.flushEvents();
+            RMTHelper.log("Waiting for the RM to detect the down node");
+            RMNodeEvent nodeEvent = RMTHelper.waitForNodeEvent(RMEventType.NODE_STATE_CHANGED,
+                                                               freeNodeUrl,
+                                                               NODE_STATE_CHANGED_TIMEOUT,
+                                                               monitor);
+
+            assertEquals(NodeState.DOWN, nodeEvent.getNodeState());
+
+            RMTHelper.log("Waiting for the RM to redeploy the down node");
+            int totalWaitingTime = 0;
+            while (this.resourceManager.getState().getFreeNodesNumber() != ONE_NODE_PER_HOST) {
+                Thread.sleep(ACTIVE_WAITING_PERIOD);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2925)